drivers/crypto/bflb-seceng.c

```diff
@@ -0,0 +1,840 @@
+#define REG_SECENG_CTRL_PROT_RD_GMAC_ID0_EN_RD      BIT(10)
+#define REG_SECENG_CTRL_PROT_RD_CDET_ID1_EN_RD      BIT(9)
+#define REG_SECENG_CTRL_PROT_RD_CDET_ID0_EN_RD      BIT(8)
+#define REG_SECENG_CTRL_PROT_RD_PKA_ID1_EN_RD       BIT(7)
+#define REG_SECENG_CTRL_PROT_RD_PKA_ID0_EN_RD       BIT(6)
+#define REG_SECENG_CTRL_PROT_RD_TRNG_ID1_EN_RD      BIT(5)
+#define REG_SECENG_CTRL_PROT_RD_TRNG_ID0_EN_RD      BIT(4)
+#define REG_SECENG_CTRL_PROT_RD_AES_ID1_EN_RD       BIT(3)
+#define REG_SECENG_CTRL_PROT_RD_AES_ID0_EN_RD       BIT(2)
+#define REG_SECENG_CTRL_PROT_RD_SHA_ID1_EN_RD       BIT(1)
+#define REG_SECENG_CTRL_PROT_RD_SHA_ID0_EN_RD       BIT(0)
+
+struct bflb_seceng {
+	struct mutex lock;
+	struct device *dev;
+	void __iomem *base;
+	struct regmap *map;
+	unsigned int initialised;
+
+	struct hwrng hwrng;
+	char hwrng_name[64];
+};
+
+struct bflb_seceng_ctx {
+	struct bflb_seceng *seceng;
+};
+
+struct regmap_config bflb_seceng_regmap_config = {
+	.reg_bits = 32,
+	.val_bits = 32,
+	.reg_stride = 4,
+	.cache_type = REGCACHE_FLAT,
+	.max_register = 512 * sizeof(u32),
+	.num_reg_defaults_raw = 512,
+	.use_relaxed_mmio = true,
+	.use_raw_spinlock = true,
+};
+
+static struct bflb_seceng *bflb_seceng_dev;
+
+static void bflb_seceng_trng_wait_ready(void)
+{
+	while (readl_relaxed(bflb_seceng_dev->base + REG_SECENG_TRNG_0_CTRL_0) &
+			REG_SECENG_TRNG_0_CTRL_0_BUSY) {
+		dev_dbg(bflb_seceng_dev->dev, "Waiting for TRNG ready");
+		schedule();
+	}
+}
+
+static void bflb_seceng_trng_init(void)
+{
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_EN | REG_SECENG_TRNG_0_CTRL_0_INT_CLR_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_EN, 1) |
+				FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_INT_CLR_1T, 1)
+	);
+
+	bflb_seceng_trng_wait_ready();
+}
+
+static void bflb_seceng_trng_refresh(void)
+{
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_DOUT_CLR_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_DOUT_CLR_1T, 1)
+	); //Clear DOUT
+
+	bflb_seceng_trng_wait_ready();
+
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_DOUT_CLR_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_DOUT_CLR_1T, 0)
+	); //Reset clear DOUT
+
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_TRIG_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_TRIG_1T, 1)
+	); //Force TRNG refresh
+
+	bflb_seceng_trng_wait_ready();
+
+	regmap_update_bits
+	(
+		bflb_seceng_dev->map,
+		REG_SECENG_TRNG_0_CTRL_0,
+		REG_SECENG_TRNG_0_CTRL_0_INT_CLR_1T,
+		FIELD_PREP(REG_SECENG_TRNG_0_CTRL_0_INT_CLR_1T, 1)
+	); //Clear INT
+
+	dev_dbg(bflb_seceng_dev->dev, "Refreshed TRNG");
+}
+
+static unsigned int bflb_seceng_read_rng_dout(struct bflb_seceng *seceng, unsigned int doutreg)
+{
+	unsigned int selectedreg = 0;
+
+	switch (doutreg) {
+	case 0:
+		selectedreg = REG_SECENG_TRNG_0_DOUT_0;
```

Comment:
These registers are all sequential, so this switch/lookup is unecessary, just replace with a direct readl_relaxed() and compute the offset like this:

readl_relaxed(seceng->base + REG_SECENG_TRNG_0_DOUT_0 + offset*4);